controller/DeliveryMgmt.php

```diff
@@ -0,0 +1,157 @@
+ * @package taoDelivery
+ */
+class DeliveryMgmt extends \tao_actions_RdfController
+{
+    use OntologyAwareTrait;
+    use TaskLogActionTrait;
+
+    /**
+     * Manage the delivery wizard form
+     *
+     * Render a creation form if form has not been submitted
+     * If has been validated use WizardFormFactory to enqueue the delivery creation
+     *
+     * @return mixed
+     * @throws \common_Exception
+     * @throws \common_ext_ExtensionException
+     */
+    public function wizard()
+    {
+        $this->defaultData();
+        $formOptions = [
+            'class' => $this->getCurrentClass(),
+        ];
+
+        $noAvailableTest = $noAvailableProvider = false;
+
+        try {
+            try {
+                $compiledDeliveryForm = (new WizardForm($formOptions))->getForm();
+
+                if ($compiledDeliveryForm->isSubmited() && $compiledDeliveryForm->isValid()) {
+                    $test = $this->getResource($compiledDeliveryForm->getValue('test'));
+                    $deliveryClass = $this->getClass($compiledDeliveryForm->getValue('classUri'));
+                    /** @var DeliveryFactory $deliveryFactory */
+                    $deliveryFactory = $this->getServiceLocator()->get(DeliveryFactory::SERVICE_ID);
+                    $initialProperties = $deliveryFactory->getInitialPropertiesFromArray($compiledDeliveryForm->getValues());
+                    return $this->returnTaskJson(CompileDelivery::createTask($test, $deliveryClass, $initialProperties));
+                }
+
+                $this->setData('compiled-delivery-form', $compiledDeliveryForm->render());
+            } catch (NoTestsException $e) {
+                $noAvailableTest = true;
+            }
+
+            try {
+                $ltiDeliveryForm = (new LtiWizardForm($formOptions))->getForm();
+
+                if ($ltiDeliveryForm->isSubmited() && $ltiDeliveryForm->isValid()) {
+                    $ltiProvider = $this->getResource($ltiDeliveryForm->getValue('ltiProvider'));
+                    $ltiPath = $ltiDeliveryForm->getValue('ltiPathElt');
+                    $deliveryClass = $this->getClass($ltiDeliveryForm->getValue('classUri'));
+
+                    /** @var LtiDeliveryFactory $deliveryFactory */
+                    $deliveryFactory = $this->getServiceLocator()->get(LtiDeliveryFactory::class);
+                    return $this->returnTaskJson(
+                        $deliveryFactory->deferredCreate($deliveryClass, $ltiProvider, $ltiPath)
+                    );
+                }
+
+                $this->setData('lti-delivery-form', $ltiDeliveryForm->render());
+            } catch (NoLtiProviderException $e) {
+                $noAvailableProvider = true;
+            }
+
+            if ($noAvailableProvider === true && $noAvailableTest === true) {
+                throw new \Exception('Delivery creation impossible without qti test neither lti provider.');
+            }
+
+
+        } catch (\Exception $e) {
```

Comment:
Does this catch something else than Exception thrown at line 104? Nested `try`'s aren't helping here... ;-)

Reply:
yes it can, but it is more a defensive approach